src/NonfungiblePositionManager.sol

```diff
@@ -0,0 +1,226 @@
+                // TODO: Burn will be coupled with decrease.
+                (uint256 tokenId) = abi.decode(params[i], (uint256));
+                burn(tokenId, sender);
+            } else {
+                revert UnsupportedAction();
+            }
+        }
+    }
+
+    /// @param param is an encoding of uint256 tokenId, uint256 liquidity, bytes hookData
+    /// @param sender the msg.sender, set by the `modifyLiquidities` function before the `unlockCallback`. Using msg.sender directly inside
+    /// the _unlockCallback will be the pool manager.
+    /// @return returns an encoding of the BalanceDelta applied by this increase call, including credited fees.
+    /// @dev Calling increase with 0 liquidity will credit the caller with any underlying fees of the position
+    function _increase(bytes memory param, address sender) internal returns (bytes memory) {
+        (uint256 tokenId, uint256 liquidity, bytes memory hookData) = abi.decode(param, (uint256, uint256, bytes));
+
+        _requireApprovedOrOwner(tokenId, sender);
+
+        // Note: The tokenId is used as the salt for this position, so every minted liquidity has unique storage in the pool manager.
+        (BalanceDelta delta,) = _modifyLiquidity(tokenRange[tokenId], liquidity.toInt256(), bytes32(tokenId), hookData);
+        return abi.encode(delta);
+    }
+
+    /// @param params is an encoding of uint256 tokenId, uint256 liquidity, bytes hookData
+    /// @param sender the msg.sender, set by the `modifyLiquidities` function before the `unlockCallback`. Using msg.sender directly inside
+    /// the _unlockCallback will be the pool manager.
+    /// @return returns an encoding of the BalanceDelta applied by this increase call, including credited fees.
+    /// @dev Calling decrease with 0 liquidity will credit the caller with any underlying fees of the position
+    function _decrease(bytes memory params, address sender) internal returns (bytes memory) {
+        (uint256 tokenId, uint256 liquidity, bytes memory hookData) = abi.decode(params, (uint256, uint256, bytes));
+
+        _requireApprovedOrOwner(tokenId, sender);
+
+        // Note: the tokenId is used as the salt.
+        (BalanceDelta delta,) =
+            _modifyLiquidity(tokenRange[tokenId], -(liquidity.toInt256()), bytes32(tokenId), hookData);
+        return abi.encode(delta);
+    }
+
+    function _mint(bytes memory param) internal returns (bytes memory) {
+        (LiquidityRange memory range, uint256 liquidity, address owner, bytes memory hookData) =
+            abi.decode(param, (LiquidityRange, uint256, address, bytes));
+
+        // mint receipt token
+        uint256 tokenId;
+        unchecked {
+            tokenId = nextTokenId++;
+        }
+        _mint(owner, tokenId);
+
+        (BalanceDelta delta,) = _modifyLiquidity(range, liquidity.toInt256(), bytes32(tokenId), hookData);
+
+        tokenRange[tokenId] = range;
+
+        return abi.encode(delta);
```

Comment:
Can we add the tokenId in the returns?

Reply:
we originally took it out so the return types for the major operations are consistent. was there a particular reason why you need it? 

our reasoning was if you wanted to mint and use the tokenId (i.e. mint & approve/permit via multicall), you would still need to preemptively derive the tokenId

Reply:
Do you mind remind me what would be the best way to track what positions a user (address) own? I am mostly thinking how would the frontend would be able to know all the tokenIds associated with an address.

> you would still need to preemptively derive the tokenId
Does it mean the caller would want to query the `nextTokenId()`?

Reply:
Probably through the mint event https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC721.sol#L169